src/partial-evaluators/IfStatement.js

```diff
@@ -13,12 +13,103 @@ import type { BabelNodeIfStatement, BabelNodeStatement } from "babel-types";
+import { Reference } from "../environment.js";
+import { joinEffects, UpdateEmpty } from "../methods/index.js";
+import { AbstractValue, Value } from "../values/index.js";
+import { construct_empty_effects } from "../realm.js";
+
+import * as t from "babel-types";
+import invariant from "../invariant.js";
 
 export default function (
   ast: BabelNodeIfStatement, strictCode: boolean, env: LexicalEnvironment, realm: Realm
-): [AbruptCompletion | Value, BabelNodeIfStatement, Array<BabelNodeStatement>] {
-  let result = env.evaluateCompletionDeref(ast, strictCode);
-  return [result, ast, []];
+): [Completion | Value, BabelNodeStatement, Array<BabelNodeStatement>] {
+  let [exprValue, exprAst, exprIO] = env.partiallyEvaluateCompletionDeref(ast.test, strictCode);
+  if (exprValue instanceof AbruptCompletion)
+    return [exprValue, t.expressionStatement((exprAst: any)), exprIO];
+  let completion;
+  if (exprValue instanceof NormalCompletion) {
+    completion = exprValue;
+    exprValue = completion.value;
+  }
+  invariant(exprValue instanceof Value);
+
+  if (!exprValue.mightNotBeTrue()) {
+    // 3.a. Let stmtCompletion be the result of evaluating the first Statement
+    let [stmtCompletion, stmtAst, stmtIO] = env.partiallyEvaluateCompletionDeref(ast.consequent, strictCode);
+
+    // 5. Return Completion(UpdateEmpty(stmtCompletion, undefined)
+    stmtCompletion = UpdateEmpty(realm, stmtCompletion, realm.intrinsics.undefined);
+    return [stmtCompletion, (stmtAst: any), exprIO.concat(stmtIO)];
+  } else if (!exprValue.mightNotBeFalse()) {
+    let stmtCompletion, stmtAst, stmtIO;
+    if (ast.alternate)
+      // 4.a. Let stmtCompletion be the result of evaluating the second Statement
+      [stmtCompletion, stmtAst, stmtIO] = env.partiallyEvaluateCompletionDeref(ast.alternate, strictCode);
+    else {
+      // 3 (of the if only statement). Return NormalCompletion(undefined)
+      stmtCompletion = realm.intrinsics.undefined;
+      stmtAst = t.emptyStatement();
+      stmtIO = [];
+    }
+    // 5. Return Completion(UpdateEmpty(stmtCompletion, undefined)
+    stmtCompletion = UpdateEmpty(realm, stmtCompletion, realm.intrinsics.undefined);
+    return [stmtCompletion, (stmtAst: any), exprIO.concat(stmtIO)];
+  }
+  invariant(exprValue instanceof AbstractValue);
+
+  // Evaluate consequent and alternate in sandboxes and get their effects.
+  realm.captureEffects();
+  let [conCompl, conAst, conIO] =
+    env.partiallyEvaluateCompletionDeref(ast.consequent, strictCode);
+  let consequentEffects = realm.getCapturedEffects();
+  invariant(consequentEffects !== undefined);
+  let [compl1, gen1, bindings1, properties1, createdObj1] = consequentEffects;
+  compl1;
+  realm.stopEffectCaptureAndUndoEffects();
+  let consequentAst = (conAst: any);
+  if (conIO.length > 0)
+    consequentAst = t.blockStatement(conIO.concat(consequentAst));
+
+  let altCompl, altAst, altIO;
```

Comment:
Is it possible to make a local function that does the work of creating the effects and ast so that we don't have to copy this code?